### ISSUE TITLE
Comparison against functions is almost always erroneous, should warn

### ISSUE THREAD

**Issue by llvmbot** [llvm/llvm-project#18274]
|  |  |
| --- | --- |
| Bugzilla Link | [17900](https://llvm.org/bz17900) |
| Version | trunk |
| OS | All |
| Reporter | LLVM Bugzilla Contributor |
| CC | @dwblaikie,@lhames |

## Extended Description 
Consider this code:

#include <strings.h>

int indx;

int isIndxZero()
{
    return index == 0;
}

Here the user has mistakenly written “indx” as “index“, which happens to be a POSIX.1  function. In this comparison, index decays to a pointer-to-function, is then compared to zero, which becomes interpreted as the null pointer constant.

As the address of a callable function, index would *never* be null, so at the very least, this is a comparison that is likely always false. 

It would be nice if at the very least we could warn that the comparison is false, or better yet, warn about comparing a function with an integer.

Also, if as a result we ended up warning for &indx == 0, that'd be great too, since that too is always false.

**Comment by Quuxplusone** [comment]
Looks like this has been fixed since Clang 3.5.
```
<source>:7:8: warning: comparison of function 'index' equal to a null pointer is always false [-Wtautological-pointer-compare]
return index == 0;
       ^~~~~    ~
<source>:7:8: note: prefix with the address-of operator to silence this warning
return index == 0;
       ^
       &
```
